napari/_qt/containers/qt_tree_model.py

```diff
@@ -0,0 +1,214 @@
+import logging
+import pickle
+from typing import List, Optional, Tuple, TypeVar
+
+from qtpy.QtCore import QMimeData, QModelIndex, Qt
+
+from ...utils.tree import Group, Node
+from ._base_item_model import _BaseEventedItemModel
+
+logger = logging.getLogger(__name__)
+NodeType = TypeVar("NodeType", bound=Node)
+NodeMIMEType = "application/x-tree-node"
+
+
+class QtNodeTreeModel(_BaseEventedItemModel[NodeType]):
+    """A QItemModel for a tree of ``Node`` and ``Group`` objects.
+
+    Designed to work with :class:`napari.utils.tree.Group` and
+    :class:`~napari._qt.containers.QtNodeTreeView`.
+
+    See docstring of :class:`_BaseEventedItemModel` and
+    :class:`~napari._qt.containers.QtNodeTreeView` for additional background.
+    """
+
+    _root: Group[NodeType]
+
+    # ########## Reimplemented Public Qt Functions ##################
+
+    def data(self, index: QModelIndex, role: Qt.ItemDataRole):
+        """Return data stored under ``role`` for the item at ``index``.
+
+        A given class:`QModelIndex` can store multiple types of data, each with
+        its own "ItemDataRole".
+        """
+        item = self.getItem(index)
+        if role == Qt.DisplayRole:
+            return item._node_name()
+        if role == Qt.UserRole:
+            return self.getItem(index)
+        return None
+
+    def index(
+        self, row: int, column: int = 0, parent: QModelIndex = QModelIndex()
+    ) -> QModelIndex:
+        """Return a QModelIndex for item at `row`, `column` and `parent`."""
+
+        # NOTE: self.createIndex(row, col, object) will create a model index
+        # that *stores* a pointer to the object, which can be retrieved later
+        # with index.internalPointer().  That's convenient and performant, but
+        # it comes with a bug if integers are in the list, because
+        # `createIndex` is overloaded and `self.createIndex(row, col, <int>)`
+        # will assume that the third argument *is* the id of the object (not
+        # the object itself).  This will then cause a segfault if
+        # `index.internalPointer()` is used later.
+
+        # XXX: discuss
```

Comment:
Does this need discussion :-) I'll be honest i don't have any real insight here, so maybe just go with which you think is best, but leave the comment documenting the other ones if we want to revisit?

Reply:
hah :) yeah, we don't _really_ need to discuss it.   it took me a while to figure out what was happening when I wrote a sample test using an EventedList of integers, so I wanted to flag it carefully.  I really doubt it's going to come up, and working around it will cost us in performance...  I think I'll just make a new issue for it so we can track it in case it ever comes up.  (see also: https://www.riverbankcomputing.com/pipermail/pyqt/2017-June/039284.html)

Reply:
ok, maybe in a follow up you can edit the comment to indicate that. i think it's fine to leave for now